[BEAM-3701][BEAM-3700] ensure pipeline options setup is using contextual classloader and not defaulting to beam-sdk classloader

rmannibucau opened the pull request:
PipelineOptions* assume the classloader is the one they got loaded with instead of relying on the TCCL.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand:
   - [ ] What the pull request does
   - [ ] Why it does it
   - [ ] How it does it
   - [ ] Why this approach
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

lukecwik commented on sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java:
```diff
@@ -620,7 +620,7 @@ private static void initializeRegistry() {
     if (!COMBINED_CACHE.containsKey(combinedPipelineOptionsInterfaces)) {
       @SuppressWarnings("unchecked")
       Class<T> allProxyClass =
-          (Class<T>) Proxy.getProxyClass(PipelineOptionsFactory.class.getClassLoader(),
```
Please add `ReflectHelpers.findClassLoader` for `Class.forName` inside:
* `findCallersClassName`
* `printHelpUsageAndExitIfNeeded`

I would appreciate if the tests were added with this PR.

rmannibucau commented:
Added tests, fixed the commit of Luke (thanks!) and added a *public* resetCache() method to ensure runners can at least control the leaks of the pipelineoptionsfactory.

A more fancy solution can be a SPI to control it but this one is a clear win-win which avoid a headache on the way to plug in the SPI (we cant reuse the TCCL for this particular one but also means it can not fit the need so having a global handler is a good trade-off).

lukecwik commented on sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java:
```diff
@@ -546,41 +517,17 @@ public Registration(Class<T> proxyClass, List<PropertyDescriptor> beanInfo) {
-        register(klass);
-      }
-    }
+  public static synchronized void resetCache() {
```
Mark this `@Experimental` so if we think of a better way we can change it.

Also, why is that unloading the classloader that contains the PipelineOptionsFactory.class not a way to solve the memory leak since your already swapping class loaders.

lukecwik commented on sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java:
```diff
@@ -1579,7 +1538,7 @@ public int compare(Method o1, Method o2) {
       Class<T> klass, ListMultimap<String, String> options, boolean strictParsing) {
     Map<String, Method> propertyNamesToGetters = Maps.newHashMap();
-    PipelineOptionsFactory.validateWellFormed(klass, REGISTERED_OPTIONS);
+    CACHE.get().validateWellFormed(klass);
```
You should move all the private static methods that refer to the inner contents of the cache into the Cache so that your working over a consistent set of values contained in the cache otherwise your not threadsafe since the contents of the CACHE may change from this line down to line 1581 where we get the CACHE again if resetCache is called in between.

Thus effectively making a non static version of the PipelineOptionsFactory.

rmannibucau replied:
Tried to refactor a bit what was possible - not public - and without impacting the whole codebase in this PR (can require a big refactoring to have a single singleton in the whole codebase but this would impact a lot of classes and would break all other PR, I think short terms we are ok with current version).

lukecwik replied:
Thinking about this more and seeing how the code has progressed, it really seems like we should make PipelineOptionsFactory not a singleton and something that can be constructed and passed around. This would solve the issue where we are leaking stuff because users could construct/pass around a PipelineOptionsFactory.

All the existing static methods would become non-static. We would store a DEFAULT instance within the PipelineOptionsFactory and create new static methods that interact with the DEFAULT instance. This also seems like it would be a lot less work to mark all the existing methods to be non-static as well then to magically call the cache in a bunch of static methods.

rmannibucau commented:
Not sure where your comment is @lukecwik (guess github ate it somehow) about having the factory not caching anything but here is my thoughts on that: I fully agree but also think it can lead to regressions and too much side effects to do it for next release so I would do it this way for now and try to refactor it for next iteration (2.5). Issues I can see is current parsing etc is slow - @iemejia said it multiple times - so we can have some perf regression if we do it regularly. I don't have the bandwidth (time) to do the full test right now but if one of you can help I'm happy to revise this opinion and try to validate we don't have regressions to do it for 2.4. Only "?" is: can we break the static utilities (get*) in 2.4?

lukecwik commented:
Parsing of what is slow (args/service loader/json/...)? I haven't seen @iemejia mention it so I don't know the context.

I think it will be a lot less work to make PipelineOptionsFactory non-static by marking the current static methods like PipelineOptionsFactory.as/create/register non-static and adding new static methods (for backwards compatibility) that just operate on a "default" PipelineOptionsFactory instance. Then your clearing logic would just work on resetting this "default" instance or better yet we wouldn't need to expose the clearing logic since in multi-classloader setups we should really be using/passing around the PipelineOptionsFactory instance or unloading the static class.

WDYT?
If you really don't have time I'll try to check that the thread safety of the cache is correct as is.

rmannibucau commented:
I'm not sure. If we have a default one then we are kind of broken and don't know what we run against + we need to use the 2 instance everywhere. Current state is backward compatible and thread safety is the same as today or said otherwise delegated to the runner (so in current state is it as safe as before since nobody calls the reset).

If we can break the API for the 2.4 we can drop all the static utilities but it sounds like some test fixes work to do, not sure I can make it for the 2.4 through this path.

lukecwik commented on sdks/java/core/src/test/java/org/apache/beam/sdk/testing/InterceptingUrlClassLoader.java:
```diff
@@ -17,22 +17,32 @@
+    }
 
     public InterceptingUrlClassLoader(final ClassLoader parent, final String... ownedClasses) {
+        this(parent, Predicates.in(Sets.newHashSet(ownedClasses)));
```
You don't need to have the private constructor for the guava predicate if you just pass in the method reference to `apply` here.

lukecwik commented on sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java:
```diff
@@ -546,113 +518,24 @@ public Registration(Class<T> proxyClass, List<PropertyDescriptor> beanInfo) {
 
   /**
    * Resets the set of interfaces registered with this factory to the default state.
+   * IMPORTANT: this is marked as experimental cause the safety usage of
```
`IMPORTANT: this is marked as experimental cause the safety usage of this method is delegated to the caller.` -> `<p>IMPORTANT: this is marked as experimental because the correct usage of this method requires appropriate synchronization beyond the scope of this method.`

lukecwik commented on sdks/java/core/src/test/java/org/apache/beam/sdk/options/PipelineOptionsFactoryTest.java:
```diff
@@ -1790,4 +1795,35 @@ public void serialize(JacksonIncompatible jacksonIncompatible, JsonGenerator jso
     }
   }
 
+  /** An options class to for the following test, it will be loaded in its own classloader. */
```
`An options class to for the following test, it will be loaded in its own classloader.` -> `Used to test that the thread context class loader is used when creating proxies.`

rmannibucau commented:
addressed

lukecwik commented:
Rerun Java Gradle PreCommit

lukecwik commented:
Run Java Gradle PreCommit

kennknowles commented:
This change appears to have broken Spark ValidatesRunner.

rmannibucau commented:
@kennknowles I'm having a look right now
